### PR TITLE
Trajectory msg proposal

### PR DESCRIPTION
**Pull request by Stifael**
# Summary
A general message that provides information about the desired kinematics over some time horizon. Any avoidance- or trajectory-module can exploit this message to pass trajectory
information to the autopilot. 

[rendered](https://github.com/Stifael/rfcs/blob/trajectory-msg/text/0004-avoidance_input_output.md)

### CODE REVIEW

**Comment by korigod** [comment]
I think it would be convenient if you add a link to the rendered RFC to your comment (https://github.com/Stifael/rfcs/blob/trajectory-msg/text/0004-avoidance_input_output.md).

File: text/0004-avoidance_input_output.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+The interface does not require to have all 5 `Anchor-Points` active. For instance, a quadratic bezier only requires three `Anchor-Points`, a cubic bezier 4 `Anchor-Points` and a simple velocity setpoint
+just the velocity part of the first `Anchor-Point`. The number of active `Anchor-Points` is defined by the field `point_valid`, which is an array of 5 elements. Each index corresponds to one `Anchor-Point` (order is preserved), where
+`0` means invalid and  `1` means valid. Any element of an active `Anchor-Point`, which is set to `NAN`, is considered as unused. For instance, a simple velocity setpoint has only the first `Anchor-Point` valid and the indices `[3-5]` set to
+a finite number with all other indices set to `NAN`. 
+
+# Unresolved Questions
+
+The number of 5 `Anchor-Points` is a suggestion. The main reason for choosing 5 is because from a computational feasibility point of view, spline parametrization with more than 4 parameters is not very common for the simple reason that it is hard to solve. 5 `Anchor-Point` can be mapped
+to a range of different spline parametrization.
+
```
Comment (thomasgubler) [inline text/0004-avoidance_input_output.md]: Could we list the pro and contra for going with variable length?

File: text/0004-avoidance_input_output.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+The trajectory message should comply with following requirements:
+* support a range of trajectory representations
+* a well-defined base message that can be extended
+* serves as input to and output from any offboard module
+
+The core of the message are five kinematic points in space, which we call `Anchor-Points`. Each `Anchor-Point` consists of an array of 11 elements. The `type`-field defines the interpretation of the 11 indices that 
+applies to all 5 `Anchor-Points`. 
+The first two suggested types are:
+* WAYPOINTS: Each `Anchor-Point` corresponds to a waypoint.
+  *	`[0-2]`: position
```
Comment (thomasgubler) [inline text/0004-avoidance_input_output.md]: is this local frame only or do we need to include the MAV_FRAME enum?

Comment (Stifael) [reply text/0004-avoidance_input_output.md]: hm, good question. I think to be able to support global  the type would need to change from float to double, which can be an overkill...

File: text/0004-avoidance_input_output.md (hunk 1 above)
Comment (Stifael) [reply text/0004-avoidance_input_output.md]: What do you mean with variable length? 
I mean the length is variable in a sense that `Anchor-Points` from 1 to 5 is supported.

Comment (thomasgubler) [reply text/0004-avoidance_input_output.md]: I mean possible more than 5

Comment (Stifael) [reply text/0004-avoidance_input_output.md]: I am not aware that mavlink supports variable length...

Comment (thomasgubler) [reply text/0004-avoidance_input_output.md]: mavlink 2.0 should do, right? I have to admit I never used it myself. Are there any caveats I am not aware of?

Comment (thomasgubler) [reply text/0004-avoidance_input_output.md]: discussed offline and dev call. Not possible with current mavlink implementation, that just clips zeros at the end. Current conclusion is to use extended fields if more points are needed

File: text/0004-avoidance_input_output.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  * `[3-5]`: velocity
+  * `[6-8]`: acceleration
+  * `[9]`  : yaw
+  * `[10]` : yaw-speed
+* BEZIER: `Anchor-Points` serve as spline parameters.
+  * `[0-2]`:  Bezier point
+  * `[3]`: Time horizon
```
Comment (thomasgubler) [inline text/0004-avoidance_input_output.md]: https://github.com/mavlink/mavlink/pull/856/files has yaw and yawrate fields here which I think make sense

FYI: @Stifael @ChristophTobler

**Comment by Stifael** [comment]
comment from @mhkabir 
>  We should be able to represent full-state waypoints (potentially even with jerk/snap components) with the trajectory message eventually.

As explained in the RFC, there are more than just one way to represent a trajectory which both are supported with this message. The more sophisticated implementation is to use the 5 points for trajectory parametrization. Doing so, the spline contains information up to snap:
- 3 points => information up to constant acceleration
- 4 points => information up to constant jerk
- 5 points => information up to constant snap

Remaining questions are now the following:
- with 6 points even snap could vary along a spline. 
  - Benefit: you have another level of freedom
  - Disadvantage: message gets bigger. The controller also needs to be able to do something smart with this information, otherwise it is useless. 
- using the 5 points as waypoints (not parametrization: see RFC description), is there any advantage of incorporating jerk/snap?
  - since a waypoint is just one state in space, I don't know how to benefit from additional jerk/snap information. 

> We should also have the other dimensions (roll, pitch and rates) in the enum.

Not sure about that. The trajectory should contain as much information about the desired newton motion, and not about orientation. Yaw is an exception and maybe a bad name and should better be called heading. Desired roll/pitch setpoints are outputs of the position controller, and therefore it does not make sense to send both. If roll/pitch information is provided as well, then you basically can skip the position controller completely which makes the message obsolete.

**Comment by mhkabir** [comment]
@Stifael 

> Not sure about that. The trajectory should contain as much information about the desired newton motion, and not about orientation. Yaw is an exception and maybe a bad name and should better be called heading. Desired roll/pitch setpoints are outputs of the position controller, and therefore it does not make sense to send both. If roll/pitch information is provided as well, then you basically can skip the position controller completely which makes the message obsolete.

If you want to specify highly dynamic trajectories which are statically unstable, you *do* have to specify orientation constraints. For example, for flying through a window which is tilted at an angle of 30 degrees, you need to provide a roll constraint of 30 degrees to that waypoint. A very nice demo from ADRL ETH :
https://youtu.be/Y7-1CBqs4x4?t=2m

This is from M. Neunert et al., "Fast nonlinear Model Predictive Control for unified trajectory optimization and tracking," 2016 IEEE International Conference on Robotics and Automation (ICRA), Stockholm, 2016, pp. 1398-1404.

**Comment by Stifael** [comment]
@mhkabir 
the way ADRL and other labs implemented MPC is quite different to the setup we have. I don't think that they send an entire trajectory to the autopilot. 
The setup that they use confirms probably with other MPC research related setups (probably similar to IDSC where I did my thesis, also around MPC):
- Offboard position/velocity controller where all the expensive computation happens. The output of the controller is is a desired attitude and scalar thrust. The MPC constraints are things like thrust and tilt limits. The optimization requires some convex optimization toolbox (I used cvxgen https://cvxgen.com/docs/index.html). There are not many useful open source optimization tool box available though.
- Onboard controller for attitude tracking. The onboard controller runs very fast up to 1000Hz (that what IDSC does).

The purpose of this trajectory message, however, is to send an entire trajectory to the autopilot.

File: text/0004-avoidance_input_output.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
+  * Start date: 2018-03-26
+  * Contributors: Dennis Mannhart <dennis@px4.io>, Martina Rivizzigno <martina@px4.io>, Christoph Tobler <christoph@px.4io>
```
Comment (mrpollo) [inline text/0004-avoidance_input_output.md]: Christoph email has the wrong domain `christoph@px.4io`

**Comment by LorenzMeier** [comment]
@Stifael Wouldn't we need heading in here for trajectory planners since the FOV is quite significant for those?

**Comment by mrivi** [comment]
@LorenzMeier  we do have the `yaw` field to specify the heading

**Comment by Stifael** [comment]
> @LorenzMeier we do have the yaw field to specify the heading

but it is not obvious from the rfc what the fields represent nor what the heading should be in between the end-points. I willl try to clarify that

**Comment by LorenzMeier** [comment]
Looking forward to the edit.

**Pull request merged**